sdk/search/Azure.Search/src/Serialization/JsonExtensions.cs

```diff
@@ -0,0 +1,473 @@
+                case JsonTokenType.False:
+                case JsonTokenType.None:
+                case JsonTokenType.Null:
+                    return ReadPrimitiveValue(ref reader);
+                case JsonTokenType.StartObject:
+                    // TODO: XXXXX - Unify on an Azure.Core spatial type
+
+                    // Return a complex object
+                    IDictionary<string, object> obj = new Dictionary<string, object>();
+                    reader.Expects(JsonTokenType.StartObject);
+                    while (reader.Read())
+                    {
+                        if (reader.TokenType == JsonTokenType.EndObject) { break; }
+                        string property = reader.ExpectsPropertyName();
+                        object value = ReadObject(ref reader);
+                        obj[property] = value;
+                    }
+                    return obj;
+                case JsonTokenType.StartArray:
+                    var list = new List<object>();
+                    while (reader.Read())
+                    {
+                        if (reader.TokenType == JsonTokenType.EndArray) { break; }
+                        object value = ReadObject(ref reader);
+                        list.Add(value);
+                    }
+                    return list.ToArray();
+                default:
+                    throw new JsonException();
+            }
+        }
+
+        /// <summary>
+        /// Serialize a SearchDocument as JSON.
+        /// </summary>
+        /// <param name="writer">JSON writer.</param>
+        /// <param name="document">The document.</param>
+        /// <param name="options">Serialization options.</param>
+        public static void WriteSearchDocument(
+            Utf8JsonWriter writer,
+            SearchDocument document,
+            JsonSerializerOptions options)
+        {
+            Debug.Assert(writer != null);
+            Debug.Assert(document != null);
+            Debug.Assert(options != null);
+
+            writer.WriteStartObject();
+            foreach (string key in document.Keys)
+            {
+                writer.WritePropertyName(key);
+                object value = document[key];
+
+                // Write the value using JsonSerializer so all of our
+                // converters take effect
+                JsonSerializer.Serialize(
+                    writer,
+                    value,
+                    value?.GetType() ?? typeof(object),
+                    options);
+            }
+            writer.WriteEndObject();
+        }
+
+        /// <summary>
+        /// Verify that the next token to be read matches our expectation or
+        /// throw an exception otherwise.
+        /// </summary>
+        /// <param name="reader">The JSON reader.</param>
+        /// <param name="expected">The expected token type.</param>
+        public static void Expects(
+            this ref Utf8JsonReader reader,
```

Comment:
I think you can use `in` when reader is not mutated in a method.

Reply:
Changed